flytekit/core/workflow.py

```diff
@@ -277,87 +320,322 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
+        if python_type is None:
+            if type(p) == list or type(p) == dict:
+                raise FlyteValidationException(
+                    f"If not a straight up promise, wf output {output_name} needs to be given an explicit type."
+                )
+            python_type = p.ref.node.flyte_entity.python_interface.outputs[p.var]
+            logger.debug(f"Inferring python type for wf output {output_name} from Promise provided {python_type}")
 
-            if (expected_outputs > 1 and len(result) == expected_outputs) or (
-                expected_outputs == 1 and result is not None
-            ):
-                if isinstance(result, Promise):
-                    v = [v for k, v in self._native_interface.outputs.items()][0]
-                    return TypeEngine.to_python_value(ctx, result.val, v)
-                else:
-                    for prom in result:
-                        if not isinstance(prom, Promise):
-                            raise Exception("should be promises")
-                        native_list = [
-                            TypeEngine.to_python_value(ctx, promise.val, self._native_interface.outputs[promise.var])
-                            for promise in result
-                        ]
-                        return tuple(native_list)
+        flyte_type = TypeEngine.to_literal_type(python_type=python_type)
 
-            raise ValueError("expected outputs and actual outputs do not match")
+        ctx = FlyteContext.current_context()
+        if ctx.compilation_state is not None:
+            raise Exception("Can't already be compiling")
+        with ctx.new_context(compilation_state=self.compilation_state) as ctx:
+            b = binding_from_python_std(
+                ctx, output_name, expected_literal_type=flyte_type, t_value=p, t_value_type=python_type
+            )
+            self._output_bindings.append(b)
+            self._python_interface = self._python_interface.with_outputs(extra_outputs={output_name: python_type})
+            self._interface = transform_interface_to_typed_interface(self._python_interface)
+
+    def ready(self) -> bool:
+        """
+        This function returns whether or not the workflow is in a ready state, which means
+          * Has at least one node
+          * All workflow inputs are bound
+
+        These conditions assume that all nodes and workflow i/o changes were done with the functions above, which
+        do additional checking.
+        """
+        if len(self.compilation_state.nodes) == 0:
+            return False
+
+        if len(self._unbound_inputs) > 0:
+            return False
+
+        return True
+
+
+class PythonFunctionWorkflow(WorkflowBase, ClassStorageTaskResolver):
+    """
+    When you assign a name to a node.
+
+    * Any upstream node that is not assigned, recursively assign
+    * When you get the call to the constructor, keep in mind there may be duplicate nodes, because they all should
```

Comment:
what does this mean?

Reply:
It means this is a months old bad comment that I wrote myself and needs to be deleted.  It's only showing up now because I made a base class.  Will remove and open a separate PR to add a proper description.